lmd/datarow.go

```diff
@@ -746,3 +746,184 @@ func (d *DataRow) WriteJSON(json *jsoniter.Stream, columns *[]*Column) {
+		hostObj, ok := p.Tables["hosts"].Index[host]
+		contactsColumn := p.Tables["hosts"].GetColumn("contacts")
+		// Make sure the host we found is actually valid
+		if ok {
+			contacts = append(contacts, *hostObj.GetStringList(contactsColumn)...)
+		}
+
+	}
+
+	// get contacts on services
+	if len(service) > 0 {
+		var serviceID strings.Builder
+		serviceID.WriteString(host)
+		serviceID.WriteString(";")
+		serviceID.WriteString(service)
+		serviceObj, ok := p.Tables["services"].Index[serviceID.String()]
+		contactsColumn := p.Tables["services"].GetColumn("contacts")
+		if ok {
+			contacts = append(contacts, *serviceObj.GetStringList(contactsColumn)...)
+		}
+	}
+
+	for _, contact := range contacts {
+		if contact == authUser {
+			canView = true
+			return
+		}
+	}
+
+	return
+}
+
+func (d *DataRow) isAuthorizedForGroup(authUser string, hostgroup string, servicegroup string) (canView bool) {
+	// Just use reflect for now, will be refactored when merging with upstream
+	// that has fixed typing.
+	var members []interface{}
+	p := d.DataStore.Peer
+	canView = false
+
+	if hostgroup != "" {
+		hostgroupObj, ok := p.Tables["hostgroups"].Index[hostgroup]
+		membersColumn := p.Tables["hostgroups"].GetColumn("members")
+		if ok {
+			// As the format for members on hostgroups differs from
+			// that of servicegroups, we manually add the items here.
+			stringMembers := *(*hostgroupObj).GetStringList(membersColumn)
+			for _, member := range stringMembers {
+				members = append(members, member)
+			}
+		}
+	} else if servicegroup != "" {
+		servicegroupObj, ok := p.Tables["servicegroups"].Index[servicegroup]
+		membersColumn := p.Tables["servicegroups"].GetColumn("members")
+		if ok {
+			members = (*servicegroupObj).GetInterfaceList(membersColumn)
+		}
+	}
+
+	for i := 0; i < len(members); i++ {
+		var hostName string
+		var serviceDescription string
+		if hostgroup != "" {
+			hostName = members[i].(string)
+		} else {
+			member := members[i].([]interface{})
+			hostName = member[0].(string)
+			serviceDescription = member[1].(string)
+		}
+		/* If GroupAuthorization is loose, we just need to find the contact
+		 * in any hosts in the group, then we can return.
+		 * If it is strict, the user must be a contact on every single host.
+		 * Therefore we return false if a contact doesn't exists on a host
+		 * and then on the last iteration return true if the contact is a contact
+		 * on the final host
+		 */
+		switch p.LocalConfig.GroupAuthorization {
+		case loose:
+			if d.isAuthorizedFor(authUser, hostName, serviceDescription) {
+				canView = true
+				return
+			}
+		case strict:
+			if !d.isAuthorizedFor(authUser, hostName, serviceDescription) {
+				canView = false
+				return
+			} else if i == len(members)-1 {
+				canView = true
+				return
+			}
+		}
+	}
+
+	return
+}
+
+func (d *DataRow) checkAuth(authUser string) (canView bool) {
+	canView = true
+	// authUser is only valid for the below tables:
+	tablesMap := map[string]struct{}{
```

Comment:
i'd make this a `map[string]bool` and since it does not change, it should not be declared on each `checkAuth()` call.

Reply:
Should I declare it as a global var instead? I guess we have no static function variables in go.

Reply:
i don't think its neccessary at all, there is a switch/case following already which sets all available cases.